cmd/hashstate.cpp

```diff
@@ -0,0 +1,311 @@
+        } else if (operation == HashStorage) {
+            // Storage
+            if (mdb_key.mv_size == kAddressLength) {
+                rc = source_table->get_next(&mdb_key, &mdb_data);
+                continue;
+            }
+            Bytes new_key(kHashLength*2+db::kIncarnationLength, '\0');
+            std::memcpy(&new_key[0], keccak256(mdb_key_as_bytes.substr(0, kAddressLength)).bytes, kHashLength);
+            std::memcpy(&new_key[kHashLength], &mdb_key_as_bytes[kAddressLength], db::kIncarnationLength);
+            std::memcpy(&new_key[kHashLength + db::kIncarnationLength], keccak256(mdb_key_as_bytes.substr(kAddressLength + db::kIncarnationLength)).bytes, kHashLength);
+            etl::Entry entry{new_key, mdb_value_as_bytes};
+            collector.collect(entry);
+            rc = source_table->get_next(&mdb_key, &mdb_data);
+        } else {
+            // Code
+            if (mdb_key.mv_size != kAddressLength+db::kIncarnationLength) {
+                rc = source_table->get_next(&mdb_key, &mdb_data);
+                continue;
+            }
+            Bytes new_key(kHashLength+db::kIncarnationLength, '\0');            
+            std::memcpy(&new_key[0], keccak256(mdb_key_as_bytes.substr(0, kAddressLength)).bytes, kHashLength);
+            std::memcpy(&new_key[kHashLength], &mdb_key_as_bytes[kAddressLength], db::kIncarnationLength);
+            etl::Entry entry{new_key, mdb_value_as_bytes};
+            collector.collect(entry);
+            rc = source_table->get_next(&mdb_key, &mdb_data);
+        }
+    }
+
+    if (rc && rc != MDB_NOTFOUND) { /* MDB_NOTFOUND is not actually an error rather eof */
+        lmdb::err_handler(rc);
+    }
+
+    SILKWORM_LOG(LogInfo) << "Started Loading" << std::endl;
+
+    switch (operation) {
+        case HashAccount:
+                collector.load(txn->open(db::table::kHashedAccounts, MDB_CREATE).get(), nullptr, MDB_APPEND, /* log_every_percent = */ 10);
+                break;
+        case HashStorage:
+                collector.load(txn->open(db::table::kHashedStorage, MDB_CREATE).get(), nullptr, MDB_APPENDDUP, /* log_every_percent = */ 10);
+                break;
+        default:
+                collector.load(txn->open(db::table::kContractCode, MDB_CREATE).get(), nullptr, MDB_APPEND, /* log_every_percent = */ 10);
+                break;
+    }
+}
+/*
+    * Extract the incarnation from an encoded account object without fully decoding it.
+*/
+uint64_t extract_incarnation(ByteView encoded) {
+    uint8_t field_set = encoded[0];
+    size_t pos{1};
+
+    if (field_set & 1) {
+        pos += encoded[pos++];
+    }
+    if (field_set & 2) {
+        pos += encoded[pos++];
+    }
+    if (field_set & 4) {
+        // Incarnation has been found.
+        uint8_t len = encoded[pos++];
+        auto [incarnation, err]{rlp::read_uint64(encoded.substr(pos, len))};
+        if (err != rlp::DecodingResult::kOk) {
+            throw err; 
```

Comment:
err is an enum ... not an exception.

Reply:
that's how those exceptions are handled everywhere in silkworm, check access_db_layer.cpp in function check_err_rlp, what being throw is err.

Reply:
cc @yperbasis for thoughts
Agree and that's why we shouldn't do that imho.
A thrown exception can be catched by a try-catch block by inheritance through its propagation in the stack.
This example will leave the exception unhandled
```
enum class [[nodiscard]] DecodingResult{
    kOk = 0,
    kOverflow,
    kLeadingZero,
    kInputTooShort,
    kNonCanonicalSingleByte,
    kNonCanonicalSize,
    kUnexpectedLength,
    kUnexpectedString,
    kUnexpectedList,
    kListLengthMismatch,
    kUnsupportedEip2718Type,
};

try {
   throw kUnexpectedLength;
} catch (const std::exception& ex) {
  // Whatever to recover or log the error
}
```
cause `kUnexpectedLength` does not inherit from std::exception.

It could only work if the catch is modified as follows
```
try {
   throw kUnexpectedLength;
} catch (...) {
  // Whatever to recover or log the error
  // but we actually don't know what exactly happened
}
```
or 
```
try {
   throw kUnexpectedLength;
} catch (const DecodingResult& ex) {
  // Whatever to recover or log the error
  // but we actually need to catch a specific type
}
```

My suggestion is to always have thrown exception to inherit from std::exception cause in higher level calls we don't know which kind of exception might be thrown at lower levels (unless perfectly documented).

Reply:
@AndreaLanfranchi Yes, I agree, we should probably create something like `DecodingException` and throw it instead of `DecodingResult`. That was my bad example, sorry. I was lazy and just threw error codes in a couple of places.
I've created Issue #203 to exceptions rather than error codes. I suggest to proceed with the PR as is and rectify the issue later.